homeassistant/components/homewizard/config_flow.py

```diff
@@ -221,19 +221,19 @@ async def _async_try_connect(ip_address: str) -> str | None:
         energy_api = HomeWizardEnergy(ip_address)
 
         try:
-            await energy_api.device()
+            return await energy_api.device()
 
-        except DisabledError:
+        except DisabledError as ex:
             _LOGGER.error("API disabled, API must be enabled in the app")
-            return "api_not_enabled"
+            raise RecoverableError("api_not_enabled") from ex
```

Comment:
Thanks for making this change. I realize one odd part i suggested is that you have to rely on str() for the error code and i realize it may have been just a partial suggestion.  How about this:

```
class RecoverableError(HomeAssistantError):
    """Raised when a connection has been failed but can be retried."""

    def __init__(self, message: str, error_code; str) -> None:
        """Init RecoverableError."""
        super().__init__(message)
        self.error_code = error_code
```

The reason why I think this is even better is because when raising exceptions you can change this:
```
        except DisabledError as ex:
            _LOGGER.error("API disabled, API must be enabled in the app")
            raise RecoverableError("api_not_enabled") from ex
Write
```
into:
```
        except DisabledError as ex:
            raise RecoverableError("API disabled, API must be enabled in the app", "api_not_enabled") from ex
```

Then at the caller where it's caught it becomes:
```
            except RecoverableError as ex:
                _LOGGER.error("Error while connecting: {str(ex)}")
                return self.async_show_form(
                    step_id="reauth_confirm",
                    errors={"base": ex.error_code},
                )
```
so you have one place to log, and getting the error code vs error message is more explicit.

Reply:
Learnt something new about exceptions 🙂!

Good suggestion, thanks! Implemented this!